File: 2024/python/solutions/day_01.py
from utils import AOCSolution


class Solution(AOCSolution):
    def _parse_input(self, input):
        self._left = []
        self._right = []
        for line in input:
            left_item, right_item = line.strip().split("   ")
            self._left.append(int(left_item))
            self._right.append(int(right_item))

    def _solve_common(self):
        self._left.sort()
        self._right.sort()

    def solve_part_1(self):
        distances = []
        for left_item, right_item in zip(self._left, self._right):
            distances.append(abs(left_item - right_item))
        total_distance = sum(distances)

        return total_distance

    def solve_part_2(self):
        right_frequencies = {}
        for item in self._right:
            right_frequencies[item] = right_frequencies.get(item, 0) + 1

        distances = []
        for item in self._left:
            distances.append(item * right_frequencies.get(item, 0))
        total_distance = sum(distances)

        return total_distance

File: scala/2024/src/day01.scala
object Day01 extends App {
  val input = scala.io.Source.fromFile("../../inputs/2024/day_01.txt").mkString

  val (left, right) = parse_input(input)

  val part_1_solution = solve_part_1(left, right)
  val part_2_solution = solve_part_2(left, right)

  println(f"Part 1: $part_1_solution")
  println(f"Part 2: $part_2_solution")

  def solve_part_1(left: Seq[Int], right: Seq[Int]): Int = {
    left.zip(right).map((left, right) => ((left - right).abs)).sum
  }

  def solve_part_2(left: Seq[Int], right: Seq[Int]): Int = {
    left.map((item) => item * right.count(_ == item)).sum
  }

  def parse_input(input: String): (Seq[Int], Seq[Int]) = {
    val input_split =
      input.linesIterator.map(_.split("   ").map(_.toInt)).toSeq

    val left = input_split.map(_(0)).sorted
    val right = input_split.map(_(1)).sorted

    (left, right)
  }

}
